### ISSUE TITLE
MVTec reproduction: timing, image-label convention, and RAM scope

### ISSUE THREAD

**Issue by Mikyx-1** [NickF93/MH-PatchCore#2]
Thanks for releasing the implementation and experiment configurations. I ran the three main MVTec configurations across all 15 categories at commit `fecde8b1c76463041102d49d999050ffe17a02c6` and have a few questions about reproducing Tables 3 and 14–16 of [the paper](https://arxiv.org/abs/2605.27748).

Accuracy is broadly close, and the parent-process RAM reduction is similar to the paper. The fitting/inference timings differ substantially, and I found one image-label convention difference caused by center cropping. I would appreciate clarification on the intended protocol.

### Setup and known differences

- Linux, Intel Core i5-12400F, CPU-only; `OMP_NUM_THREADS=4`, `OPENBLAS_NUM_THREADS=4`, `MKL_NUM_THREADS=4`, and `CUDA_VISIBLE_DEVICES=""`. Torch and FAISS reported four threads. The paper's MVTec setup uses an i7-9750H with 12 CPU threads available, so these are different hardware/thread conditions.
- Python 3.11.11; torch 2.10.0+cpu; torchvision 0.25.0+cpu; numpy 2.2.2; scipy 1.14.1; scikit-learn 1.8.0; faiss-cpu 1.13.2. The pinned direct requirements were installed. `weights="DEFAULT"` resolves to WideResNet50-2 ImageNet V2.
- Configurations: `configs/mvtec/baselines/mvtec_vanilla_patchcore.yaml`, `configs/mvtec/baselines/mvtec_streaming_mh_patchcore.yaml`, and `configs/mvtec/ablations/mvtec_ablation_no_mahalanobis.yaml`.
- Batch size 16, eight workers per loader, original seeds (vanilla 0; streaming 42), no training augmentation. Vanilla retains the supplied 10% coreset; streaming retains the supplied IPCA/merge-reduce settings and cap of 1,000.
- Dataset: [Voxel51/mvtec-ad](https://huggingface.co/datasets/Voxel51/mvtec-ad/tree/30a183a3b96e3aef953f230784b123b719b09d97), 3,629 training images and 1,725 test images. All 5,354 image hashes were checked against the download manifest; all 1,258 mask paths were checked and local mask hashes retained. Mirror source basenames were restored through symlinks to preserve filename ordering; image contents were unchanged.
- Each category/configuration ran sequentially in a **fresh process**, rather than all categories in one process. This can affect allocator retention and RAM peaks. Only dataset/output paths, the single-category list, and progress interval were overridden in the YAMLs.
- A wrapper invoked upstream `run_experiment(load_run_config(config_path))` with the tracked upstream source unchanged. It retained references to predictions/model and saved raw arrays after upstream profiling. A separate supervisor sampled process-tree memory once per second; the measurements include this monitoring overhead.

All nine upstream tests passed. I independently recomputed all 45 image/pixel AUROC pairs from saved scores/maps and checked their agreement with the upstream CSVs. Two runs affected by external CPU contention or system suspend were repeated; every saved numerical array was bit-identical, and the replacement runs supply the timings below. These are single accepted runs per category/configuration, without timing confidence intervals.

### 1. Inference and fitting time

Mean fit time is macro-averaged over categories; inference is image-weighted ms/image.

| Configuration | Paper mean fit, s | Local mean fit, s | Paper inference, ms/image | Local inference, ms/image |
|---|---:|---:|---:|---:|
| Vanilla | 12.444 | 207.835 | 467.092 | 393.415 |
| Euclidean streaming | 54.822 | 170.236 | 39.592 | 149.355 |
| Mahalanobis streaming | 57.913 | 199.997 | 42.610 | 151.574 |

This gives 2.60× local inference speedup for Mahalanobis versus 10.96× from the published table. The different hardware/thread settings prevent treating that ratio difference alone as a contradiction of the paper.

Our inference measurement uses the upstream timer around `infer_dataloader(test_loader)`: loader startup, loading/preprocessing, backbone, retrieval, and map generation are included; metric calculation and visualization are separate. No additional warmup was introduced.

- What exact command, thread settings, warmup procedure and timing boundaries produced the published inference results? Were backbone extraction and test-loader startup included there as well?
- Does the reported 12.444-second vanilla fitting mean include both feature extraction and the supplied 10% greedy coreset selection? Were any cached/precomputed stages outside the reported fit timer?

### 2. Accuracy and image labels after cropping

The following uses the released code's image labels. AUROCs are percentages, macro-averaged over 15 categories.

| Configuration | Paper image / pixel | Local image / pixel |
|---|---:|---:|
| Vanilla | 99.1 / 98.0 | 99.123 / 98.020 |
| Euclidean streaming | 97.9 / 96.5 | 97.867 / 96.432 |
| Mahalanobis streaming | 98.9 / 97.8 | 98.761 / 97.778 |

For example, local Mahalanobis image AUROC on screw is 92.847% versus 93.7% in Table 3. I have not tuned settings to close these differences.

Separately, [`normalize_batch` derives tuple-batch image labels from masks](https://github.com/NickF93/MH-PatchCore/blob/fecde8b1c76463041102d49d999050ffe17a02c6/mhpc/core/batching.py#L49-L55), and [`_infer_labels_from_masks`](https://github.com/NickF93/MH-PatchCore/blob/fecde8b1c76463041102d49d999050ffe17a02c6/mhpc/core/batching.py#L121-L126) labels an image normal when its transformed mask is empty.

In the mirrored dataset, `grid/test/glue/008-97.png` is an anomalous image whose entire defect is removed by center cropping:

- Original mask: 1,024 × 1,024; 2,384 positive pixels; bounding box `(246, 0, 319, 48)` in `(left, top, right, bottom)` coordinates.
- Transformed mask: zero positive pixels, producing image label `0` instead of the folder-derived `1`.
- Image SHA-256: `f38f3c95343197328df609d0fa54c93417cb6882fb7c48e674007fd561678383`.
- Mask SHA-256: `1a830f239e297590b2d022e5f60ed553639492a664b0a8994ec5d584318ab7e6`.

Using folder-derived image labels with identical saved scores changes Mahalanobis grid AUROC from 99.838% to 98.162%, and its overall image AUROC from 98.761% to 98.649%. Vanilla's overall image AUROC changes from 99.123% to 99.005%. This is one unique image affected in all three configurations; pixel AUROC is unchanged.

- Is evaluating image labels from the visible, cropped masks intentional, and was Table 3 computed with this convention? Or should the loader preserve the original dataset image label independently of the transformed mask?

### 3. Peak RAM measurement scope

| Configuration | Paper max RAM, labeled GB | Local max parent RSS, GiB | Local max process-tree PSS, GiB |
|---|---:|---:|---:|
| Vanilla | 5.412 | 4.541 | 7.296 |
| Euclidean streaming | 2.611 | 2.457 | 7.126 |
| Mahalanobis streaming | 2.775 | 2.475 | 7.242 |

The local parent-RSS reduction is **45.5%**, broadly consistent with the paper's **48.7%** reduction. Local GiB uses powers of 1024; the paper labels its values GB.

The [released profiler](https://github.com/NickF93/MH-PatchCore/blob/fecde8b1c76463041102d49d999050ffe17a02c6/mhpc/eval/profiling.py#L250) samples parent RSS every 0.5 seconds and excludes workers. Our additional process-tree PSS measurement proportionally accounts for shared pages, rather than summing RSS. Both full-tree peaks occurred with 18 processes, including the two persistent eight-worker loaders. The resulting 0.7% reduction measures a different scope and does not contradict the parent-RSS reduction. Both are sampled peaks, and fresh category processes also differ from the default all-category process lifetime.

- Were Tables 14–15 collected with the same parent-only RSS profiler? Does the intended peak cover fitting only, inference only, or the full fit/inference/metrics/artifact sequence? Could the units and exclusion of worker processes be clarified?

### 4. Experiment provenance

Could you identify the exact commit and commands used for Tables 3 and 14–16, and share the corresponding environment freeze and raw `per_dataset.csv` / profiling CSVs if available? This would help distinguish version or protocol differences from hardware effects. This reproduction covers MVTec only; I have not evaluated the private industrial datasets.

Thank you for any clarification.

**Comment by NickF93** [comment]
Thank you for the exceptionally detailed reproduction and for separating accuracy, timing, and memory-scope observations. Your accuracy and parent-process RSS results broadly reproduce ours. The largest difference concerns runtime, so we address the protocol in detail below.

### 1. What commands and thread settings were used?

The three relevant entry points were:

```bash
python run_mhpc.py --config configs/mvtec/baselines/mvtec_vanilla_patchcore.yaml

python run_mhpc.py --config configs/mvtec/baselines/mvtec_streaming_mh_patchcore.yaml

python run_mhpc.py --config configs/mvtec/ablations/mvtec_ablation_no_mahalanobis.yaml
```

The MVTec experiments were CPU-only. `CUDA_VISIBLE_DEVICES` was configured at the system-environment level so that no CUDA device was visible; no GPU computation was used.

We did not manually set `OMP_NUM_THREADS`, `OPENBLAS_NUM_THREADS`, or `MKL_NUM_THREADS`, nor did we set CPU affinity. The workstation had 12 CPU threads available. The released implementation configures FAISS to use four threads.

Your explicit four-thread limits are therefore one difference between the environments, but we do not think they alone explain the complete timing pattern. In particular, your vanilla inference is faster than ours, while both streaming configurations are slower. The configurations have different bottlenecks: vanilla searches a substantially larger retained feature bank, while the bounded variants shift more of their relative cost toward feature extraction, preprocessing, and transformation.

The two machines also differ in CPU architecture, numerical-library builds, scheduler activity, filesystem caching, thermal and frequency behavior, and other system state. The reported timings are observed measurements on the reported system, not hardware-independent latency guarantees.

We executed each configuration multiple times using the same code and experimental settings. The experiments were run on a development workstation rather than a dedicated benchmarking system, so timings varied because of OS scheduling, background processes, cache state, and other transient conditions. No algorithmic or configuration changes were made between these executions.

During result consolidation, for each MVTec category and timing field, the lowest available value among configuration-matched executions was retained, with the intention of approximating the least-contended execution. The `MEAN` values in the timing tables are arithmetic means across the 15 MVTec categories after that consolidation; they are not means across repeated executions or estimates with confidence intervals.

This consolidation and workstation variability can contribute to the observed differences, but we cannot attribute the full discrepancy—especially the large vanilla fitting difference—to one specific factor.

### 2. What exactly is included in fitting time?

The fitting timer wraps the complete:

```python
model.fit(train_loader)
```

operation.

It includes:

- Initial iteration and startup of the training loader.
- Image loading and preprocessing.
- Backbone feature extraction.
- Feature reduction and transformation where configured.
- Memory aggregation and materialization.
- Fitting the nearest-neighbour scorer.

For vanilla PatchCore, this includes feature extraction and the complete configured 10% greedy coreset selection. No feature or coreset stage was cached or precomputed outside the fit timer.

DataLoader construction, model construction, and pretrained-weight loading occur before the fit timer and are excluded.

### 3. What exactly is included in inference time?

No explicit inference warmup pass was performed.

The inference timer wraps the complete:

```python
model.infer_dataloader(test_loader)
```

operation.

It includes:

- The first test-loader iteration, including worker startup where applicable.
- Test-image loading and preprocessing.
- Backbone feature extraction.
- Configured feature transformations.
- Nearest-neighbour retrieval.
- Image and patch scoring.
- Anomaly-map generation.

Metric computation and visualization or artifact generation occur afterward in separately profiled phases and are excluded from inference time.

For each category, per-image inference time is calculated as the complete category inference duration divided by its number of test images. The final `MEAN` is the arithmetic mean of the resulting 15 category-level values after consolidation.

### 4. Why does the reproduced speedup differ?

Your Mahalanobis streaming result gives a 2.60× inference speedup over vanilla, compared with 10.96× in our table.

We consider your measurement valid for your reported system and execution environment. The speedup ratio combines two workloads with different bottlenecks, so it can vary substantially across CPUs, thread policies, FAISS and PyTorch builds, and system state. The fact that your vanilla inference is faster while your streaming inference is slower also shows that a simple overall machine-speed factor cannot explain the difference.

We therefore interpret the published timing comparison as system-specific rather than as an expected invariant across hardware.

### 5. Is the cropped-mask image-label convention intentional?

Yes. The evaluation intentionally concerns anomalies visible in the model’s transformed input.

The image and mask are jointly resized to 256 pixels and center-cropped to \(224 \times 224\). For MVTec tuple batches, the image-level label is then inferred from the transformed mask. If the complete defect lies outside the retained crop, the evaluated image is consequently treated as normal.

Table 3 was computed using this convention. This is also why the evaluated `grid` split contains 56 anomalous labels rather than the 57 labels obtained directly from the original test folders.

We also confirmed that the image and mask hashes you supplied match the corresponding files used in our experiments. Your observed label change is therefore an accurate description of the intended visible-region protocol.

### 6. What is the RAM measurement scope?

Tables 14–15 use parent-process RSS only. DataLoader worker processes are excluded.

RSS is sampled every 0.5 seconds. For each category, profiling begins immediately before fitting and continues through:

1. Fitting.
2. Calibration.
3. Inference.
4. Metric computation.
5. Artifact generation.

Model and DataLoader construction before profiling, final CSV writing, and cleanup after profiling are excluded.

All 15 MVTec categories are processed sequentially in the same process. Consequently, allocator retention from earlier categories can affect later category measurements. Running every category in a fresh process, as in your reproduction, is a different process-lifetime protocol.

Each category’s RAM value is its maximum sampled parent RSS. The global maximum is the largest of those 15 category peaks, while reported mean values are means across the category peaks.

Your process-tree PSS result includes the persistent DataLoader workers and answers a different resource-accounting question. It therefore does not contradict the parent-RSS reduction. Your measured 45.5% parent-RSS reduction is broadly consistent with our reported 48.7%.

### 7. Are the RAM values GB or GiB?

The profiler divides RSS bytes by \(1024^2\), producing MiB, and the table-generation step divides those values by 1024. The displayed values are therefore GiB.

The currently published arXiv version labels these values as GB. This has been corrected to GiB in the manuscript source, but the correction has not yet been uploaded to arXiv. The numerical values themselves are unchanged.

### 8. What implementation and environment correspond to the tables?

The public commit:

```text
fecde8b1c76463041102d49d999050ffe17a02c6
```

together with the three supplied YAML configurations represents the released base implementation and experiment settings.

The paper records the main execution environment as Linux, Intel Core i7-9750H, 12 CPU threads available, 15 GiB RAM, 2 GiB swap, Python 3.11.11, and PyTorch 2.10.0. The public dependency files provide additional environment information.

We do not have a confirmed run-specific full environment freeze beyond the versions recorded in the paper and the released dependency files. In particular, `environment-linux-64.yml` should not be interpreted as a confirmed exact export of the MVTec execution environment.

The per-dataset and profiling CSVs are not currently included in the public release, so we cannot provide a public link to them at present.

Overall, your results reproduce the principal accuracy and parent-process memory findings. The runtime discrepancy, particularly for vanilla fitting, is not fully explained by the information currently available. We will therefore keep this issue open while we investigate the timing behavior further, including stage-level costs and sensitivity to threading, library builds, and system state. We will report any additional findings here.
